app/temporal/src/activities.ts

```diff
@@ -81,175 +88,150 @@ export function makeActivities(
-    }): Promise<{ errorFileReference: string; errorCount: number }> => {
-      console.time("validations");
-      const referenceId = params.fileReference.split("-")[1].split(".")[0];
-      const fileData = await fileStore.getFile(
-        params.bucket,
-        params.fileReference
-      );
-      const jsonData = JSON.parse(fileData.toString());
-      const patchedData = applyPatches(jsonData, params.patches);
-      const errorData = dataAnalyzer.processDataValidations(
-        patchedData,
+      skip: number;
+      limit: number;
+    }): Promise<number> => {
+      const jsonData: DataSet = await database.mongoClient
+        .db(params.importerId)
+        .collection<DataSetRow>("data")
+        .find()
+        .skip(params.skip)
+        .limit(params.limit)
+        .toArray();
+
+      const validationResults = dataAnalyzer.processDataValidations(
+        jsonData,
         params.validatorColumns,
         params.stats
       );
-      await fileStore.putFile(
-        params.bucket,
-        params.outputFileReference,
-        Buffer.from(JSON.stringify(errorData))
-      );
-      console.timeEnd("validations");
-      return {
-        errorFileReference: params.outputFileReference,
-        errorCount: errorData.length,
-      };
+
+      for (const validationResult of validationResults) {
+        await database.mongoClient
+          .db(params.importerId)
+          .collection("data")
+          .updateOne(
+            { __sourceRowId: validationResult.rowId },
+            {
+              $push: {
+                [`data.${validationResult.column}.messages`]:
+                  validationResult.messages,
+              },
+            }
+          );
+      }
+
+      return validationResults.length;
+    },
+    applyPatches: async (params: {
+      importerId: string;
+      patches: DataSetPatch[];
+    }): Promise<void> => {
+      for (const patch of params.patches) {
+        const targetColumn = `data.${patch.column}`;
+        await database.mongoClient
+          .db(params.importerId)
+          .collection("data")
+          .updateOne(
+            { __sourceRowId: patch.rowId },
+            { $set: { [targetColumn]: patch.newValue } }
+          );
+      }
     },
     generateStats: async (params: {
-      bucket: string;
-      fileReference: string;
+      importerId: string;
       uniqueColumns: string[];
-      patches: DataSetPatch[];
     }): Promise<SourceFileStatsPerColumn> => {
       console.time("generate-stats");
-      const fileData = await fileStore.getFile(
-        params.bucket,
-        params.fileReference
-      );
-      const jsonData: DataSet = JSON.parse(fileData.toString());
-      const patchedData = applyPatches(jsonData, params.patches);
-      const stats = dataAnalyzer.getStats(patchedData, params.uniqueColumns);
+      const jsonData: DataSet = await database.mongoClient
+        .db(params.importerId)
+        .collection<DataSetRow>("data")
```

Comment:
Wir können es erstmal so lassen, aber ich denke es wäre effizienter, wenn wir das in Zukunft mit einer einfachen Aggregation lösen

Reply:
ja - aber nicht ganz so einfach da die facets über die 16 MB kommen. Hier ist die Aggragation die ich zum testen genommen habe:

`db.getCollection("data").aggregate([
  {
    $facet: {
      BSNRStats: [
        {
          $group: {
            _id: "$data.BSNR",
            count: { $sum: 1 }
          }
        },
        {
          $project: {
            k: "$_id",
            v: {
              nonunique: "$count"
            }
          }
        }
      ],
      NameKundeStats: [
        {
          $group: {
            _id: "$data.Name Kunde",
            count: { $sum: 1 }
          }
        },
        {
          $project: {
            k: "$_id",
            v: {
              nonunique: "$count"
            }
          }
        }
      ]
    }
  },
  {
    $project: {
      stats: {
        $arrayToObject: {
          $concatArrays: ["$BSNRStats", "$NameKundeStats"]
        }
      }
    }
  }
])
`

![image](https://github.com/Jank1310/datadolphin/assets/12624982/d6c2acff-0b79-42cc-97be-7d2e576955a9)

wir lassen das jetzt erstmal so - können wir ja dann noch optimieren

Reply:
$match to the rescue :-)
Habs hinbekommen:

`db.getCollection("data").aggregate([
    {
        $facet: {
            BSNRStats: [
                {
                    $group: {
                        _id: "$data.BSNR.value",
                        count: { $sum: 1 }
                    }
                },
                {
                    $match: {
                        count: { $ne: 1 }
                    }
                },
                {
                    $project: {
                        _id: 0,
                        k: "$_id",
                        v: "$count"
                    }
                }
            ],
            NameKundeStats: [
                {
                    $group: {
                        _id: "$data.Name Kunde.value",
                        count: { $sum: 1 }
                    }
                },
                {
                    $match: {
                        count: { $ne: 1 }
                    }
                },
                {
                    $project: {
                        _id: 0,
                        k: "$_id",
                        v: "$count"
                    }
                }
            ]
        }
    },
    {
        $replaceRoot: {
            newRoot: {
                BSNR: {
                    nonuniqe: {
                        $arrayToObject: "$BSNRStats"
                    }
                }",
                "Name Kunb": {
                    nonuniqe: {
                        $arrayToObject: "$NameKundeStats"
                    }
                }
            }
        }
    }
])
`